litmus-portal/graphql-server/utils/misc.go

```diff
@@ -31,55 +31,53 @@ func RandomString(n int) string {
+	)
+
 	file, err := os.Open(template)
 	if err != nil {
 		return []byte{}, err
 	}
 	defer file.Close()
 	scanner := bufio.NewScanner(file)
-	var lines []string
 
-	if !*cluster.AgentNsExists || !*cluster.AgentSaExists {
-		comment := fmt.Sprintf("# This is an auto-generated file. DO NOT EDIT")
-		lines = append(lines, comment)
+	if cluster.AgentNsExists == nil {
+		cluster.AgentNsExists = &defaultState
 	}
-	if !*cluster.AgentNsExists {
-		var nsYaml string
-		if cluster.AgentNamespace != nil {
-			nsYaml = fmt.Sprintf("apiVersion: v1\nkind: Namespace\nmetadata:\n  name: %s\n---", *cluster.AgentNamespace)
-		} else {
-			nsYaml = fmt.Sprintf("apiVersion: v1\nkind: Namespace\nmetadata:\n  name: litmus\n---")
-		}
-		lines = append(lines, nsYaml)
+	if cluster.AgentSaExists == nil {
+		cluster.AgentSaExists = &defaultState
 	}
-	if !*cluster.AgentSaExists {
-		var saYaml string
-		if cluster.Serviceaccount != nil && cluster.AgentNamespace != nil {
-			saYaml = fmt.Sprintf("apiVersion: v1\nkind: ServiceAccount\nmetadata:\n  name: %s\n  namespace: %s\n---", *cluster.Serviceaccount, *cluster.AgentNamespace)
-		} else if cluster.Serviceaccount == nil && cluster.AgentNamespace != nil {
-			saYaml = fmt.Sprintf("apiVersion: v1\nkind: ServiceAccount\nmetadata:\n  name: litmus\n  namespace: %s\n---", *cluster.AgentNamespace)
-		} else if cluster.Serviceaccount != nil && cluster.AgentNamespace == nil {
-			saYaml = fmt.Sprintf("apiVersion: v1\nkind: ServiceAccount\nmetadata:\n  name: %s\n  namespace: litmus\n---", *cluster.Serviceaccount)
-		} else {
-			saYaml = fmt.Sprintf("apiVersion: v1\nkind: ServiceAccount\nmetadata:\n  name: litmus\n  namespace: litmus\n---")
-		}
-		lines = append(lines, saYaml)
+
+	if !*cluster.AgentNsExists && cluster.AgentNamespace != nil && *cluster.AgentNamespace != "" {
+		AgentNamespace = *cluster.AgentNamespace
+	} else {
+		AgentNamespace = DefaultAgentNamespace
 	}
 
+	if !*cluster.AgentSaExists && cluster.Serviceaccount != nil && *cluster.Serviceaccount != "" {
+		ServiceAccountName = *cluster.Serviceaccount
+	} else {
+		ServiceAccountName = DefaultServiceAccountName
+	}
+
+	var (
+		headerStr         = "# This is an auto-generated file. DO NOT EDIT\n"
+		namspaceStr       = "----\napiVersion: v1\nkind: Namespace\nmetadata:\n  name: " + AgentNamespace + "\n"
+		serviceAccountStr = "---\napiVersion: v1\nkind: ServiceAccount\nmetadata:\n  name: " + ServiceAccountName + "\n  namespace: " + AgentNamespace + "\n---"
+	)
+
+	saYaml := fmt.Sprintf(headerStr + namspaceStr + serviceAccountStr)
+	lines = append(lines, saYaml)
+
 	for scanner.Scan() {
 		line := scanner.Text()
-		if strings.Contains(line, "# This is an auto-generated file. DO NOT EDIT") {
-			if !*cluster.AgentNsExists || !*cluster.AgentSaExists {
-				line = strings.Replace(line, "# This is an auto-generated file. DO NOT EDIT", "", -1)
-			}
-		} else if strings.Contains(line, "#{CID}") {
+		if strings.Contains(line, "#{CID}") {
```

Comment:
We need to remove the string - `# This is an auto-generated file. DO NOT EDIT` from the manifest incase the sa and ns yaml get appended.

Reply:
Anyway, we need this string permanently because the whole manifest is autogenerated.

Reply:
If it's okay to have this string twice in the manifests, the PR looks good.

Reply:
this string will appear a single time in all the cases.

Reply:
What about this string? - https://github.com/litmuschaos/litmus/blob/23b1957258526dc387f4e8a45ea2e11f9e7d78d1/litmus-portal/graphql-server/manifests/cluster-subscriber.yml#L1

Reply:
Oh, Okay, we can remove it from there